amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java

```diff
@@ -484,19 +484,64 @@ static boolean checkIfDescendantAndAddNewLeasesForAncestors(final String shardId
                 // because we'd have done it when creating a lease for this shard.
             } else {
                 final Shard shard = shardIdToShardMapOfAllKinesisShards.get(shardId);
-                final Set<String> parentShardIds = getParentShardIds(shard, shardIdToShardMapOfAllKinesisShards);
 
+                final Set<String> childShardIds = shardIdToChildShardIdsMap.get(shardId);
+                if (childShardIds != null && childShardIds.size() == 2) {
```

Comment:
For the case below, what will happen when we reach 5 though the path 11,4,5 but we have leases for 13 and 14. Would be end up creating leases for 10? Can we add test cases for these scenarios?

     * 0 1 2 3 4   5- shards till epoch 102
     * \ / \ / |   |
     *  6   7  4   5- shards from epoch 103 - 205
     *   \ /   |  /\
     *    8    4 9  10 - shards from epoch 206 (open - no ending sequenceNumber)
     *          /\  /\
     *        11 12 13 14

Reply:
We wouldn't create a lease for 10 in this case because at the time shard 5 is being traversed (top of the stack), neither of its children will have leases so we won't hit this condition.

If we had a lease for 9, we still wouldn't create a lease for 10 because we would have a complete hash range. Will add test cases for these.

Reply:
Discussed offline. If we have leases for 13 and 14 but traverse the left half of the tree (from 11 -> 9 -> 5) before traversing the right half of the tree, we will have no way of knowing a priori at which level the hash range is complete. This breaks the TRIM/TIMESTAMP case.

Assume n = 0 is the root (shard 5). The child shards optimization in the previous revision only solves a special case where the hash range is partially complete at n = 1 (shard 10). For all n > 1, since we don't know what the other half of the tree looks like, we will go all the way back to the beginning and create a lease for shard-5. This could lead to out-of-order processing since we already have leases for 13 and 14, but will then be retroactively creating leases again for 5 (and then 10 when 5 reaches SHARD_END). 

However, this is current with the existing behavior when reading from TRIM/TIMESTAMP so there is no regression. However we should scope this work in a followup PR to modify the algorithm and fix this bug.